crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -0,0 +1,364 @@
+        for function in ssa.functions.values_mut() {
+            self.defunctionalize(function);
+        }
+    }
+
+    /// Defunctionalize a single function
+    fn defunctionalize(&mut self, func: &mut Function) {
+        let mut target_function_ids = HashSet::new();
+
+        for block_id in func.reachable_blocks() {
+            let block = &func.dfg[block_id];
+            let instructions = block.instructions().to_vec();
+
+            for instruction_id in instructions {
+                let instruction = func.dfg[instruction_id].clone();
+                let mut replacement_instruction = None;
+                // Operate on call instructions
+                let (target_func_id, arguments) = match instruction {
+                    Instruction::Call { func: target_func_id, arguments } => {
+                        (target_func_id, arguments)
+                    }
+                    _ => continue,
+                };
+
+                match func.dfg[target_func_id] {
+                    // If the target is a function used as value
+                    Value::Param { .. } | Value::Instruction { .. } => {
+                        // Collect the argument types
+                        let argument_types: Vec<Type> =
+                            arguments.iter().map(|arg| func.dfg.type_of_value(*arg)).collect();
+
+                        // Collect the result types
+                        let result_types =
+                            vecmap(func.dfg.instruction_results(instruction_id), |result| {
+                                func.dfg.type_of_value(*result)
+                            });
+                        // Find the correct apply function
+                        let apply_function = self.get_apply_function(&FunctionSignature {
+                            parameters: argument_types,
+                            returns: result_types,
+                            runtime: func.runtime(),
+                        });
+                        target_function_ids.insert(apply_function.id);
+
+                        // Replace the instruction with a call to apply
+                        let apply_function_value_id = func.dfg.import_function(apply_function.id);
+                        if apply_function.dispatches_to_multiple_functions {
+                            let mut new_arguments = vec![target_func_id];
+                            new_arguments.extend(arguments);
+                            replacement_instruction = Some(Instruction::Call {
+                                func: apply_function_value_id,
+                                arguments: new_arguments,
+                            });
+                        } else {
+                            replacement_instruction = Some(Instruction::Call {
+                                func: apply_function_value_id,
+                                arguments,
+                            });
+                        }
```

Comment:
```suggestion
                        if apply_function.dispatches_to_multiple_functions {
                            arguments.insert(0, target_func_id);
                        }
                        let func = apply_function_value_id;
                        replacement_instruction = Some(Instruction::Call { func, arguments });
```
I opted for `insert` over creating a new Vec and extending/appending to it with the old Vec since we expect most functions to have relatively few parameters so the cost of moving over a few parameters is probably smaller than allocating a new Vec.